### PR TITLE
🐛 : FIX : Long labels overflowing in extended navigation rail.

### LINKED ISSUE #110901: [adaptive_scaffold] `NavigationDestination` with icon + label overflows within navigation rail.
Screen Shot 2022-09-02 at 11 50 12 PM

**Expected results:** The AppBar sizes to the same width as the largest element in the NavigationListView

**Actual results:**  Overflow Error

## Steps to Reproduce

<details>
<summary>Code sample</summary>

1. Clone [flutter/packages](https://github.com/flutter/packages) into the folder of your choice (at the time of writing, `AdaptiveScaffold` has not been released.
2. Add adaptive_scaffold to the flutter project:
```yaml
dependencies
    adaptive_scaffold:
        path: <path_to_the_packages_repo>/packages/adaptive_scaffold/
```
3. Add the following to`main.dart`

```dart
import 'package:flutter/material.dart';
import 'package:adaptive_scaffold/adaptive_scaffold.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: AdaptiveScaffold(
        destinations: [
          NavigationDestination(icon: Icon(Icons.home), label: "ABCDEFGHIJKLMNOP"),
          NavigationDestination(icon: Icon(Icons.home), label: "ABCDEFGHIJKLMNOP")
        ],
      ),
    );
  }
}
```

</details>

To be clear, this wouldn't be a problem if the `label:` property allowed a widget. If that were the case, any user could specify the way that the widget renders whenever there was an overflow error. Right now, the `label:` property is a string, which makes it impossible to account for a long `NavigationDestination.label`

### PR DESCRIPTION
**Pull request by aliasgar4558**
It has been observed that while in extended navigation rail, if longer labels are given the it is overflowing. Problem identified in "NavigationRail" widget only in case of extended version of it. This PR includes the fix for the same.

Fixes https://github.com/flutter/flutter/issues/110901.

*NO breaking changes.*

## Pre-launch Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Comment by aliasgar4558** [comment]
Closed by mistake, re-opened the PR.

**Comment by dkwingsmt** [comment]
When I run it locally, with you additional `Expanded`, the `navigation_rail_test` throws tons of errors for multiple tests. And even after I revert your changes (removing the `Expanded`), your test throws the following assertion:
```
'package:flutter/src/material/navigation_rail.dart': Failed assertion: line 121 pos 16: '!extended
|| (labelType == null || labelType == NavigationRailLabelType.none)': is not true.
```

Yet somehow all checks have passed. Have you run it locally? I don't know what's happening.

**Comment by aliasgar4558** [comment]
@dkwingsmt - What are the configuration that I need to make to run the tests additionally, as when I tried to run the test, it didn't allowed me to execute.

**Comment by dkwingsmt** [comment]
What do you mean it doesn't allow you to execute? To run the test you `cd` into `packages/flutter` then run
```dart
flutter test test/material/navigation_rail_test.dart
```

**Comment by aliasgar4558** [comment]
Sure, will check and update here accordingly

**Comment by aliasgar4558** [comment]
@dkwingsmt - is 'dev' channel fine to work on this one? Or shall I work on master or some other environment?

**Comment by dkwingsmt** [comment]
Any branch should be OK, since the navigation rail hasn't been changed for a while.

**Comment by aliasgar4558** [comment]
I tried with stable channel, but it shows several compile time errors which are related to conflicts into pubspec versions, so thats the purpose of asking if I can try with dev or master instead. Mind me asking, what env did you used to verify the changes?

**Comment by dkwingsmt** [comment]
I'm using the main channel. Can you try with that?

**Comment by aliasgar4558** [comment]
Yeah sure, Allow me some time. Will check out in `main` channel

**Comment by aliasgar4558** [comment]
@dkwingsmt  - Do we have to make any changes in pubspec in `flutter/packages/flutter' folder, as I am unable to execute `pub get` there.?

**Comment by aliasgar4558** [comment]
@dkwingsmt - You are right, 18 tests are breaking. However, I found one issue with my test, so will update that and commit into this PR. 

I don't feel like I shall change any other tests as it might result into something else to break. Please suggest your thoughts!

**Comment by aliasgar4558** [comment]
@dkwingsmt  - Please check if you can verify now, as I am not getting any test breaks. 

Root cause : `Expanded` widget was the one causing issue as the widget is not bounded by parent.
Remedy : Replaced with `Flexible`

**Comment by dkwingsmt** [comment]
All tests now passes, but after I revert your changes (removing the `Flexible` wrap) your test still passes. Can you check it out?

**Comment by aliasgar4558** [comment]
@dkwingsmt - sure, I will add more fine grain test which checks for "Flexible" with the longer texts.

**Comment by dkwingsmt** [comment]
The idea is that the unit test you add should fail before your change and pass after your change, so that it prevents your change from being reverted in the future. One unit test is enough as long as it's on the point.

**Comment by aliasgar4558** [comment]
@dkwingsmt  - Yes I agree. Sure thing, will update the test so in future if something changes, it can be indicated. Thanks for the insights.

**Comment by dkwingsmt** [comment]
Your test checks if the text can be found within a `Flexible`. It's not testing the feature, but the implementation. What you really care is that no overflow is reported, instead of whether it's implemented via `Flexible` or any other widgets. 

And in fact, your implementation does not solve the problem. Here's what I got from running your fix:

image

How to test it:
1. Open `examples/api/lib/material/navigation_rail/navigation_rail.1.dart`
2. Change line 73 to a really long label, such as 
```dart
              destinations: const <NavigationRailDestination>[
                NavigationRailDestination(
                  icon: Icon(Icons.favorite_border),
                  selectedIcon: Icon(Icons.favorite),
                  label: Text('LongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLongLong'),
                ),
```
3. Run it, for example, 
```
flutter run ~/dev/flutter/examples/api/lib/material/navigation_rail/navigation_rail.1.dart -d chrome
```
4. Make the window really small

**Comment by aliasgar4558** [comment]
@dkwingsmt - I think problem is with how navigation rail is used in UI. It might require width  constraints in normal case scenario also. Suggest your insights please!

**Comment by aliasgar4558** [comment]
yes @dkwingsmt - There are 2 issues basically here. First, what I fixed - that is when long text added, it shall not overflow (wrapped in flexible) & second is `NavigationRail` without max width constraints is over flowing when size is compressed (what you have faced - in your case, if you wrap navigation rail with `SizedBox` & give width, then it will not overflow)

for first point, I have made the fix & it will work when navigation rail is available with max width constraint but the second thing that you faced, for that we need to take a change. [We can raise a bug for that for "NavigationRail"]

For 2nd point - I think we can introduce "maxWidth" parameter in `NavigationRail` which will help to constraint its width. [New change]

**Comment by aliasgar4558** [comment]
Allow me some time, I will send the code snippet here - which will demonstrate my fix with an example for you.

**Comment by aliasgar4558** [comment]
@dkwingsmt any update on this one mate?

**Comment by dkwingsmt** [comment]
> @dkwingsmt any update on this one mate?

Hi! I'm still waiting for the snippet you mentioned. Can you demonstrate the scenario that this PR fixes? Such as some comparison screenshots would be even better.

**Comment by aliasgar4558** [comment]
@dkwingsmt Ohh, Okay sure. I will be sending that across. Thank you

**Comment by aliasgar4558** [comment]
- This is the problem you faced when you shrink the screen size very small.
problem_with_smaller_screen

- But the problem is not with the fix that I made, it is because of "NavigationRail"'s width is not constraint & the other view is crashing/overflowing.

Example screenshot when `NavigationRail`'s width is given :
navigation_rail_with_size_constraints

- Now below is the screenshot of what I fixed. [Same width given to navigation rail & maxLines specified for longer rail destination title.]
demonstration_of_my_fix


-- So from above screenshots, it seems like `NavigationRail` shall be provided with width constraints in consumer applications so other views can be rendered correctly.

@dkwingsmt - Please have a check on this one & you can try the same thing out & let me know if anything. 
(I referred to `navigation_rail.1.dart` file for demo purpose)

**Comment by dkwingsmt** [comment]
(Sorry I missed your PR last week) Thanks for your screenshot comparison, which are super clear about what was fixed.

Now, my problem with your unit test is that, it tests an implementation detail ("the text must be wrapped within a `Flexible` widget") instead of the behavior ("the text must wrap / the max-line must be triggered"). The widget might be refactored in the future with same functionality implemented differently, and you want your test to still be valid. My suggestion is that your test can check the height of the widget in this setup and ensure it's below some number that wasn't possible before the PR, which I think is one of the most observable difference.

**Comment by aliasgar4558** [comment]
Not a problem for the delay. Thanks for the valuable insights. 

However, I didn't got your point to check what widget's height while in test. Can you please let me know more details on it ?

**Comment by aliasgar4558** [comment]
@dkwingsmt - Updated test case to check `Text` widget's height value & compared with the other smaller destination's height (which is a normal one - which can be adjusted in a single line). Hope this is what you are expecting.

**Comment by dkwingsmt** [comment]
I ran the widget in your test individually but the result is not the desired one as you demonstrated.

Before PR: (with exception)
image

After PR: 
image

Although they're indeed different, they do not demonstrate the ellipsis fix that you showed earlier.

I found out that it was because you missed the "ellipsis" setting for `Text` in your test. With the "ellipsis",

Before PR: (with exception)
image


After PR:
image

Anyway, since with either cases the before-PR version throws a layout exception, I think we can omit checking anything here.

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 1]
```diff
@@ -3631,6 +3632,96 @@ void main() {
...
+      of: find.byType(Flexible),
+      matching: find.descendant(
+        of: find.byType(Align),
+        matching: find.descendant(
+          of: find.byType(FadeTransition),
+          matching: longLabelNavDestinationFinder,
+        ),
+      ),
+    );
+    expect(flexibleWithNavDestinationText, findsOneWidget);
```
Comment (dkwingsmt) [inline packages/flutter/test/material/navigation_rail_test.dart:3722]:
```suggestion
  // If the widget manages to layout without throwing an overflow exception,
  // the test passes.
```

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 2]
```diff
@@ -3631,6 +3632,96 @@ void main() {
     expect(tester.takeException(), isNull);
   });
 
+  testWidgets('NavigationRail labels shall not overflow if longer texts provided - extended', (WidgetTester tester) async {
+
+    // Adding widget separately to ensure the navigation rail's width is fixed so navigation destination shall wrap itself as expected
```
Comment (dkwingsmt) [inline packages/flutter/test/material/navigation_rail_test.dart:3637]: I'm not sure what this comment is meant for. Maybe this?
```suggestion
    // The navigation rail has a narrow width constraint. It should wrap itself.
```

**Comment by aliasgar4558** [comment]
@dkwingsmt - Yes, ellipsis was intentionally not added in tests to demonstrate that text would be wrapped if navigation rail has a width constraint, so it can be determined that `Flexible` is making some difference.

So, Do you suggest we can exclude the test cases for this fix?

Also - Mind me asking, how do you guys run widget in testing to get the above screenshots ??

**Comment by aliasgar4558** [comment]
I've applied the changes you suggested. @dkwingsmt - Thank you for the valuable insights

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 3]
```diff
@@ -3631,6 +3632,62 @@ void main() {
...
+    final Finder normalLabelNavDestinationFinder = find.text(
+      'Abc',
+    );
+    final Finder longLabelNavDestinationFinder = find.text(
+      'Very long bookmark text for navigation destination',
+    );
+    expect(normalLabelNavDestinationFinder, findsOneWidget);
+    expect(longLabelNavDestinationFinder, findsOneWidget);
+  // If the widget manages to layout without throwing an overflow exception,
+  // the test passes.
```
Comment (dkwingsmt) [inline packages/flutter/test/material/navigation_rail_test.dart:3688]:
```suggestion
    // If the widget manages to layout without throwing an overflow exception,
    // the test passes.
```

**Review by dkwingsmt — APPROVED** [review]
LGTM

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 4]
```diff
@@ -3631,6 +3632,62 @@ void main() {
     expect(tester.takeException(), isNull);
   });
 
+  testWidgets('NavigationRail labels shall not overflow if longer texts provided - extended', (WidgetTester tester) async {
+
+    // The navigation rail has a narrow width constraint. The text should wrap.
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3637]: Consider adding a regression comment such as:
```suggestion
  testWidgets('NavigationRail labels shall not overflow if longer texts provided - extended', (WidgetTester tester) async {
    // Regression test for https://github.com/flutter/flutter/issues/110901.
    // The navigation rail has a narrow width constraint. The text should wrap.
```

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 5]
```diff
@@ -3631,6 +3632,62 @@ void main() {
...
+    await tester.pumpWidget(MaterialApp(
+      theme: ThemeData(useMaterial3: true),
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3639]: Consider removing `theme: ThemeData(useMaterial3: true),`
This is no more required (M3 is the default now).

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 6]
```diff
@@ -3631,6 +3632,62 @@ void main() {
...
+      home: Builder(
+        builder: (BuildContext context) {
+          return MediaQuery.withClampedTextScaling(
+            minScaleFactor: 1.0,
+            maxScaleFactor: 1.0,
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3644]: Consider removing `MediaQuery.withClampedTextScaling`, it is not needed for this particular test.

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 7]
```diff
@@ -3631,6 +3632,62 @@ void main() {
...
+
+    expect(find.byType(NavigationRail), findsOneWidget);
+    final Finder normalLabelNavDestinationFinder = find.text(
+      'Abc',
+    );
+    final Finder longLabelNavDestinationFinder = find.text(
+      'Very long bookmark text for navigation destination',
+    );
+    expect(normalLabelNavDestinationFinder, findsOneWidget);
+    expect(longLabelNavDestinationFinder, findsOneWidget);
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3686]: Maybe consider adding variables for the labels (for instance, `normalLabel` and `longLabel`) and using them for the repro code and for the finders:
```suggestion
    expect(find.text(normalLabel), findsOneWidget);
    expect(find.text(longLabel), findsOneWidget);
```

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 8]
```diff
@@ -3631,6 +3632,62 @@ void main() {
...
+    // If the widget manages to layout without throwing an overflow exception,
+    // the test passes.
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3688]: Consider checking `tester.takeException()`, it makes it more explicit that, before the fix, an exception was thrown.
```suggestion
    // If the widget manages to layout without throwing an overflow exception,
    // the test passes.
    expect(tester.takeException(), isNull);
```

**Review by bleroux — CHANGES_REQUESTED** [review]
Great fix!
I propose some minor changes in the test to improve the readability and the consistency with existing tests.

**Comment by bleroux** [comment]
@aliasgar4558 Sorry for the burden, you will have to rebase this PR on the `master` branch (you can close this one and file a new one if it is easier for you), `main` branch can not be targeted for PRs on flutter/flutter (the bots won't work with it).

FYI, for the other Flutter repos, main branch is ok, especially for `flutter/engine` but this is not the case for flutter/flutter.

**Comment by aliasgar4558** [comment]
@bleroux - Thanks for the suggestions. It seems good to have those in the PR so will add that. 

Also, Just to confirm - instead of `main` branch, we have to target this PR to `master` branch. Is that correct?

**Comment by bleroux** [comment]
> Also, Just to confirm - instead of `main` branch, we have to target this PR to `master` branch. Is that correct?

Yes, this is required in flutter/flutter. Sorry for the burden.

**Comment by aliasgar4558** [comment]
@bleroux - Not problem at all. I am here to learn and contribute & you guys are helping me. So all good. Will do the needful and update PR.

File: packages/flutter/test/material/navigation_rail_test.dart (hunk 5 above)
Comment (aliasgar4558) [reply packages/flutter/test/material/navigation_rail_test.dart]: Removed.

File: packages/flutter/test/material/navigation_rail_test.dart (hunk 6 above)
Comment (aliasgar4558) [reply packages/flutter/test/material/navigation_rail_test.dart]: Removed.

File: packages/flutter/test/material/navigation_rail_test.dart (hunk 7 above)
Comment (aliasgar4558) [reply packages/flutter/test/material/navigation_rail_test.dart]: Updated as per the suggestion.

**Comment by aliasgar4558** [comment]
@bleroux - I have updated test as per your suggested changes & now PR is targeting to `master` branch. Please review once you get some time.

File: packages/flutter/test/material/navigation_rail_test.dart [hunk 9]
```diff
@@ -3631,6 +3632,56 @@ void main() {
...
+                ),
+              ],
+            ),
+          );
+        },
+      ),
+    ));
+
+    const String normalLabel = 'Abc';
+    const String longLabel = 'Very long bookmark text for navigation destination';
```
Comment (bleroux) [inline packages/flutter/test/material/navigation_rail_test.dart:3675]: Great! Please move these declarations above (line 3637) and use these variables in the pumpWidget body (lines 3653 & 3658).

Comment (aliasgar4558) [reply packages/flutter/test/material/navigation_rail_test.dart]: Good catch. Updated now.

**Review by bleroux — APPROVED** [review]
LGTM! Thanks for your contribution 🙏

**Comment by aliasgar4558** [comment]
@dkwingsmt @bleroux - Thank you so much for your valuable insights for this PR. Really Appreciate it 😊

**Comment by dkwingsmt** [comment]
Thank YOU for the contribution, and continuously addressing our comments!

**Pull request merged**